Make Stroke and Fill properties non-null.

rpoelstra opened the pull request:
As the Stroke's and Fill's properties are already always non-null I defined them as such in the struct itself to prevent needless unwrapping when using the structs.
This also has the advantage that any default values are always enforced at a single point in code.

rpoelstra committed:
Make Stroke and Fill properties non-null.

Make sure the defaults are applied in a single location.

richardpiazza commented:
Thanks for the submission. I'll take a look!

richardpiazza commented:
@rpoelstra Overall, I like the direction! I'm going to make a few suggestions on how the code is structured just to bring it closer in line with my style. (When I find some more time, I'll update the readme with https://github.com/SwiftActions/StyleGuide)

richardpiazza commented on Sources/SwiftSVG/Fill.swift:
```diff
@@ -2,11 +2,15 @@ import Swift2D
     public var rule: Rule = .nonZero
     
-    public init() {}
+    public init(color: String?, opacity: Double?, rule: Rule?) {
+        self.color = color ?? "black"
+        self.opacity = opacity ?? 1
+        self.rule = rule ?? .nonZero
+    }
```
Let's move the default values to the initializer.
```suggestion
    /// Initialize aggregate data related to a `Fill`.
    ///
    /// - parameters:
    ///   - color: Name of or hexadecimal value used to color the fill. Default: "black"
    ///   - opacity: The level of opaqueness given applied to the color. Default: 1.0
    ///   - rule: Algorithm used to determine the inside of the shape. Default: `.nonZero`
    public init(
        color: String = "black", 
        opacity: Double = 1.0, 
        rule: Rule = .nonZero
    ) {
        self.color = color
        self.opacity = opacity
        self.rule = rule
    }
```

rpoelstra replied:
Oh, I forgot to mention that I didn't supply default values for the constructor to prevent the type of bug that was in your original code where you had forgotten to supply the path's fill rule to the struct's constructor. Without default values such bug is more difficult to make, as you have to supply some value.
Hence, I see various drawbacks to this kind of constructor.

richardpiazza replied:
I would like to maintain a parameterless initialization option. If the properties of the `Fill` struct are going to be non-optional, then those default values need to be present somewhere. I prefer them to be as defaults specified in the method signature rather than as _un-exposed_ values with the init method. This adds in documentation efforts well for those who are not looking directly at the code.

rpoelstra replied:
The parameterless initialisation is easy of course, it can simply call self.init(...) with whatever arguments it likes. I've already done this on the 'rpoelstra' branch in my fork of your code, where I've already added some other construct as well.
Fair enough about the unexposed values. Maybe define them as static constants on the struct to have them readily available instead of hardcoding them in various places?

richardpiazza commented on Sources/SwiftSVG/PresentationAttributes.swift:
```diff
@@ -80,10 +80,9 @@ public extension PresentationAttributes {
-            fill.color = fillColor ?? "black"
-            fill.opacity = fillOpacity ?? 1.0
-            return fill
+            return Fill(color: fillColor,
+                        opacity: fillOpacity,
+                        rule: fillRule)
```
Now with the required parameters, we'll need to still supply values when nil...
```suggestion
            return Fill(
                color: fillColor ?? "black",
                opacity: fillOpacity ?? 1.0,
                rule: fillRule ?? .nonZero
            )
```

This would also be a good opportunity to perform the "none" check you referred to...

```swift
if let color = fillColor, color.caseInsensitiveCompare("none") == .orderedSame {
    return nil
}
```

rpoelstra replied:
Having to supply the defaults at the point of calling is exactly what I tried to prevent.
As normally when any of the properties is nil you want the default value. With this change you need to remember the correct value when using the struct.

richardpiazza commented on Sources/SwiftSVG/PresentationAttributes.swift:
```diff
@@ -98,14 +97,12 @@ public extension PresentationAttributes {
-            stroke.lineJoin = strokeLineJoin ?? .miter
-            stroke.miterLimit = strokeMiterLimit
-            return stroke
+            return Stroke(color: strokeColor,
+                          width: strokeWidth,
+                          opacity: strokeOpacity,
+                          lineCap: strokeLineCap,
+                          lineJoin: strokeLineJoin,
+                          miterLimit: strokeMiterLimit)
```
Same styling as `Fill`... same 'none' check as well?

richardpiazza commented on Sources/SwiftSVG/Stroke.swift:
```diff
@@ -2,15 +2,20 @@ import Swift2D
     
-    public init() {
-        
+    public init(color: String?, width: Double?, opacity: Double?, lineCap: LineCap?, lineJoin: LineJoin?, miterLimit: Double?) {
+        self.color = color ?? "black"
+        self.width = width ?? 1
+        self.opacity = opacity ?? 1
+        self.lineCap = lineCap ?? .butt
+        self.lineJoin = lineJoin ?? .miter
+        self.miterLimit = miterLimit ?? 4
```
Let's move the default values to the initializer like `Fill`...
Why the default `miterLimit` of 4?

rpoelstra replied:
I took the miterLimit value from the SVG docs.

rpoelstra committed:
Added parameterless initialiser to Stroke and Fill.

rpoelstra committed:
Extracted out default values to own struct for Stroke and Fill.

Added documentation to initialisers.

rpoelstra commented:
Hi,

I've made some changes which resolve most of your issues I think.
-A made explicit constants for the default values, so they are easier to locate and re-useable.
-Added parameterless initialisers
-Added (mostly) your documentation.

I kept the nullable arguments in the initialisers as they prevent code-duplication at the call side.
What do you think about a structure like this?
